torch_geometric/transforms/add_positional_encoding.py

```diff
@@ -0,0 +1,151 @@
+from typing import Any, Optional
+
+import numpy as np
+import torch
+from scipy.sparse.linalg import eigs, eigsh
+from torch import Tensor
+from torch_scatter import scatter_add
+from torch_sparse import spspmm
+
+from torch_geometric.data import Data
+from torch_geometric.data.datapipes import functional_transform
+from torch_geometric.transforms import BaseTransform
+from torch_geometric.utils import (
+    get_laplacian,
+    remove_self_loops,
+    to_scipy_sparse_matrix,
+)
+from torch_geometric.utils.num_nodes import maybe_num_nodes
+
+
+def diagonal_weight(edge_index, edge_weight, num_nodes=None):
+    loop_mask = edge_index[0] == edge_index[1]
+    loop_index = edge_index[:, loop_mask][0]
+    loop_edge_attr = edge_weight[loop_mask]
+
+    N = maybe_num_nodes(edge_index, num_nodes)
+    diag_attr = torch.zeros((N, ), dtype=edge_weight.dtype)
+    diag_attr[loop_index] = loop_edge_attr
+    return diag_attr
+
+
+class BaseAddPE(BaseTransform):
```

Comment:
Not really a big fan of the base class here. For now, I suggest to make `add_pe` a function outside the class:
```python
def add_attr(data: Data, value: Any, cat: bool, attr_name: Optional[str] = None):
```
We could also think of adding this function to `base_transform.py`, and use it in other `transforms` as well, e.g., in `transforms.Cartesian`. WDYT?

Reply:
This is actually a fantastic idea. However, it might modify many files (although I have not counted yet). So, how about creating `add_attr` outside the class for now, and making it a part of `base_transform.py` later in the other PR?

Reply:
Maybe it could be part of `Data`?